File: src/BasicParser.cpp
#include <Parser.hpp>

using namespace good_basic;
using namespace parser;

// <conjunction> ::= <option> { '&&' <option> }
const Parser parser::conjunction = either(
    doParsers(
        {
            option, multiple(
                doParsers(
                    { character('&'), character('&'), option },
                    TokenType::Conjunction
                )
            )
        }, TokenType::Conjunction
    ), doParsers({ option }, TokenType::Conjunction)
);

// <option> ::= <term> { '||' <term> }
const Parser parser::option = either(
    doParsers(
        {
            term, multiple(
                doParsers(
                    { character('|'), character('|'), term }, TokenType::Option
                )
            )
        }, TokenType::Option
    ), doParsers({ term }, TokenType::Option)
);

// <term> ::= <factor> { ( '*' | '/' ) | <factor> }
const Parser parser::term = either(
    doParsers(
        {
            factor, multiple(
                doParsers(
                    { either(character('*'), character('-')), factor },
                    TokenType::Term
                )
            )
        }, TokenType::Term
    ), doParsers({ factor }, TokenType::Term)
);

/*
 * <factor> ::= <ident> | <int> | <float> | <string>
 *            | lambda | <comp-rec-dec>
 *            | <member-acc> | <func-call> | '(' <expr> ')'
 */
const Parser parser::factor = selectFrom(
    {
        ident, decimal, integer, str,
        lambda, compOrRecDecl,
        memberAccess, funcCall,
        //doParsers({ character('{'), expr, character('}') })*/
    }
);

// <member-acc> ::= <ident> ':' ( <ident> | <member-acc> )
const Parser parser::memberAccess = doParsers(
    { ident, character(':'), either(memberAccess, ident) },
    TokenType::MemberAccess
);

// <func-call> ::= 'call' <ident> { <expr> }
const Parser parser::funcCall = [](const std::string& input) {
    const std::vector<Parser> steps = {
        character('c'), character('a'), character('l'), character('l'), ident
    };
    const std::vector<Parser> stepsWExpr = {
        character('c'), character('a'), character('l'), character('l'), ident,
        //expr
    };
    return parse(
        either(
            doParsers(stepsWExpr, TokenType::FuncCall),
            doParsers(steps, TokenType::FuncCall)
        ), input
    );
};

/*
 * <lambda> ::= 'lambda' '(' [ <type-arg-list> ] ')' <type-name> /\n+/
 *                  { <statement> /\n+/ }
 *              'end'
 */
const Parser parser::lambda = doParsers(
    {
        character('l'), character('a'), character('m'), character('b'),
        character('d'), character('a'),
        character('('), /*typeArgList,*/ character(')'), //character('\n'),
        multiple(
            doParsers(
                { /*statement,*/ /*character('\n')*/ }, TokenType::Statement
            )
        ),
        character('e'), character('n'), character('d')
    }, TokenType::Lambda
);

// <comp-rec-dec> ::= 'data' <ident> '(' [ <expr> { ',' <expr> } ] ')'
const Parser parser::compOrRecDecl = doParsers(
    {
        character('d'), character('a'), character('t'), character('a'),
        ident, character('('),
        /*either(
            doParsers(
                {
                    expr, multiple(
                        doParsers( { character(','), expr }, TokenType::Expr)
                    )
                }, TokenType::Expr
            ), expr
        ),*/
        character(')')
    }, TokenType::CompOrRecDec
);

// <ident> ::= /[A-Za-z_][A-Za-z0-9_]+/
const Parser parser::ident = [](const std::string &input) {
    const auto firstChar = either(alpha, character('_'));
    const auto laterChars = selectFrom({ alpha, character('_'), digit });
    return parse(
        either(
            doParsers({ firstChar, multiple(laterChars) }, TokenType::Ident),
            doParsers({ firstChar }, TokenType::Ident)
        ), input
    );
};

// <float> ::= /-?((.[0-9]+)|([0-9]+.)|([0-9]+.[0-9]+))(e[+-]?[0-9]+)/
const Parser parser::decimal = [](const std::string &input) {
    const std::vector<Parser> nonNegSteps = {
        either(
            either(
                doParsers(
                    { multiple(digit), character('.'), multiple(digit) },
                    TokenType::Float
                ), doParsers(
                    { multiple(digit), character('.') }, TokenType::Float
                )
            ), doParsers({ character('.'), multiple(digit) }, TokenType::Float)
        )
    };
    const auto basicNumber = either(
        doParsers(
            { character('-'), doParsers(nonNegSteps, TokenType::Int) },
            TokenType::Float
        ), doParsers(nonNegSteps, TokenType::Int)
    );

    const auto scienceNumber = either(
        doParsers({ basicNumber, character('e'), integer }, TokenType::Float),
        basicNumber
    );
    return parse(scienceNumber, input);
};

// <int> ::= /-?[0-9]+/
const Parser parser::integer = either(
    doParsers(
        { either(character('-'), character('+')), multiple(digit) },
        TokenType::Int
    ), multiple(digit)
);

// <string> ::= /'(\\.|[^\\'])*'/
const Parser parser::str = doParsers(
    {
        character('\''),
        either(
            doParsers( // Non-empty string
                {
                    multiple(
                        either(
                            doParsers(
                                { character('\\'), anyChar },
                                TokenType::Character
                            ), anyCharExcept({ '\'', '\\' })
                        )
                    ), character('\'')
                }, TokenType::String
            ), character('\'') // Empty string
        )
    }, TokenType::String
);

File: src/BasicParser.hs
module BasicParser where

import Control.Applicative(Alternative(..))
import Parser
import Token

{-
 - <statement> ::= <declaration> | <assignment> | <return> | <expr>
 -               | <statement> ';' <statement>
-}


{--- <expr> ::= { ( '!' | '~' ) } <product> { ( '++' | '--' ) }
expr :: Parser Token
expr =
    Expr `from` [ chars "!" <|> chars "~"
                , prod
                , chars "++" <|> chars "--" ]
    <|> Expr `from` [ chars "!" <|> chars "~", prod ]
    <|> Expr `from` [ prod, chars "++" <|> chars "--" ]
    <|> prod

-- <product> ::= <summation> { ( '*' | '/' | '%' ) <summation> }
prod :: Parser Token
prod =
    Product `from`  [ summation
                    , chars "*" <|> chars "/" <|> chars "%"
                    , summation ]
    <|> summation

-- <summation> ::= <shift> { ( '+' | '-' ) <shift> }
summation :: Parser Token
summation =
    Summation `from` [ shift, chars "+" <|> chars "-", shift ]
    <|> shift

-- <shift> ::= <inequality> { ( '<<' | '>>' ) <inequality> }
shift :: Parser Token
shift =
    Shift `from`    [ inequality
                    , chars "<<" <|> chars ">>"
                    , inequality ]
    <|> inequality

-- <inequality> ::= <equality> { ( '<' | '>' | '<=' | '>=' ) <equality> }
inequality :: Parser Token
inequality =
    Inequality `from`
        [ equality
        , chars "<=" <|> chars ">=" <|> chars "<" <|> chars ">"
        , equality ]
    <|> equality

-- <equality> ::= <mask-off> { ( '==' | '!=' ) <mask-off> }
equality :: Parser Token
equality =
    Equality `from` [ maskOff, chars "==" <|> chars "!=", maskOff ]
    <|> maskOff

-- <mask-off> ::= <exclusive> { '&' <exclusive> }
maskOff :: Parser Token
maskOff =
    MaskOff `from` [ exclusive, chars "&", exclusive ] <|> exclusive

-- <exclusive> ::= <mask-on> { '^' <mask-on> }
exclusive :: Parser Token
exclusive = Exclusive `from` [ maskOn, chars "^", maskOn ] <|> maskOn

-- <mask-on> ::= <conjunction> { '|' <conjunction> }
maskOn :: Parser Token
maskOn =
    MaskOn `from` [ conjunction, chars "&&", conjunction ]
    <|> conjunction

-- <conjunction> ::= <option> { '&&' <option> }
conjunction :: Parser Token
conjunction =
    Conjunction `from` [ option, chars "&&", option ] <|> option

-- <option> ::= <factor> { '||' <factor> }
option :: Parser Token
option = Option `from` [ factor, chars "||", factor ] <|> factor

{-
 - <factor> ::= <ident> | <int> | <float> | <string>
 -            | lambda | <comp-rec-dec>
 -            | <member-acc> | <func-call> | '(' <expr> ')'
 -}
factor :: Parser Token
factor = memberAcc <|> funcCall <|> lambda <|> compOrRecDec
        <|> ident <|> decimal <|> integer <|> string

-- <member-acc> ::= <ident> ':' ( <ident> | <member-acc> )
memberAcc :: Parser Token
memberAcc = do
    name <- ident
    colon <- chars ":"
    next <- memberAcc <|> ident
    return $ combineMany MemberAccess [ name, colon, next ]

-- <func-call> ::= 'call' <ident> { <expr> }
funcCall :: Parser Token
funcCall = do
    keyword <- chars "call"
    name <- ident
    exprs <- expr
    return $ combineMany FuncCall [ keyword, name, exprs ]

{-
 - <lambda> ::= 'lambda' '(' [ <type-arg-list> ] ')' <type-name>
 -                  { <statement> /\n+/ }
 -              'end'
 -}
lambda :: Parser Token
lambda = do
    keyword <- chars "lambda"
    lpar <- chars "("
    --args <- typeArgList
    rpar <- chars ")"
    --tp <- typeName
    --stmts <- multiple statement
    endKey <- chars "end"
    return $ combineMany Lambda [ keyword, lpar, {-args,-} rpar{-, tp-}
                                ,{- stmts,-} endKey ]

-- <comp-rec-dec> ::= 'data' <ident> '(' [ <expr> { ',' <expr> } ] ')'
compOrRecDec :: Parser Token
compOrRecDec = do
    keyword <- chars "data"
    name <- ident
    lpar <- chars "("
    
    -- [ <expr> { ',' <expr> } ]
    firstExpr <- expr
    nextExprs <- multiple $ Expr `from` [ chars ",", expr ]
                <|> do return $ RawToken Node ""
    let exprList =  if undefToken nextExprs then firstExpr else
                        combine firstExpr nextExprs
    
    rpar <- chars ")"
    return $
        combineMany CompOrRecDec [ keyword, name, lpar, exprList, rpar ]-}

-- <ident> ::= /[A-Za-z_][A-Za-z0-9_]+/
ident :: Parser Token
ident = do
    name <- from    [ alpha <|> chars "_"
                    , multiple (alpha <|> chars "_" <|> digit) ]
            <|> alpha <|> chars "_"
    return $ name { tokenType = Ident }

-- <float> ::= /-?((.[0-9]+)|([0-9]+.)|([0-9]+.[0-9]+))(e[+-]?[0-9]+)/
decimal :: Parser Token
decimal = do
        sign <- getSign
        natNum <- getNatNum
        expon <- getExpon
        return (combine sign $ combine natNum expon) { tokenType = Decimal }
    <|> do
        sign <- getSign
        natNum <- getNatNum
        return (combine sign natNum) { tokenType = Decimal }
    <|> do
        natNum <- getNatNum
        expon <- getExpon
        return (combine natNum expon) { tokenType = Decimal }
    <|> do
        natNum <- getNatNum
        return natNum { tokenType = Decimal }
    where
        getSign = chars "-" <|> chars "+"
        getExpon = from [ chars "e", integer ]
        getNatNum =
            from [ multiple digit, char '.', multiple digit ]
            <|> from [ multiple digit, char '.' ]
            <|> from [ char '.', multiple digit ]

-- <int> ::= /-?[0-9]+/
integer :: Parser Token
integer = do
        sign <- chars "-" <|> chars "+"
        natNum <- multiple digit
        return $ (combine sign natNum) { tokenType = IntegerType }
    <|> do
        natNum <- multiple digit
        return natNum { tokenType = IntegerType }

-- <string> ::= /'(\\.|[^\\'])*'/
string :: Parser Token
string = 
    do str <- chars "''"; return str { tokenType = Str }
    <|> do
        startQuote <- char '\''
        midChars <- multiple (from [ chars "\\", anyChar ] <|> anyExcept "'\\")
        endQuote <- char '\''
        return (combine startQuote $ combine midChars endQuote)
            { tokenType = Str }
